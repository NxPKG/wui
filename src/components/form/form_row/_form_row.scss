/**
 * 1. Coerce inline form elements to behave as block-level elements.
 * 2. For inline forms, we need to add margin if the label doesn't exist.
 */
.wuiFormRow {
  display: flex; /* 1 */
  flex-direction: column; /* 1 */
  max-width: $wuiFormMaxWidth;

  + .wuiFormRow,
  + .wuiButton {
    margin-top: $wuiSize;
  }
}

.wuiFormRow--fullWidth {
  max-width: 100%;
}

.wuiFormRow--hasEmptyLabelSpace {
  margin-top: ($wuiFontSizeXS * $wuiLineHeight) + $wuiSizeXS; /* 2 */
  // the following ensure that contents that aren't inherently the same height
  // as inputs will align to the vertical center
  min-height: $wuiFormControlHeight;
  padding-bottom: 0;
  justify-content: center;
}

.wuiFormRow__labelWrapper {
  display: flex;
  flex-wrap: wrap;
  justify-content: space-between;
  margin-bottom: $wuiSizeXS;
}

.wuiFormRow--horizontal {
  flex-direction: row;
  align-items: stretch;

  .wuiFormRow__label {
    @include wuiTextBreakWord;
    hyphens: auto;
    max-width: 100%; // Fixes IE
  }

  .wuiFormRow__labelWrapper {
    display: block;
    line-height: $wuiFormControlCompressedHeight - 1px; // The 1px less helps the alignment of the text baseline
    width: calc(33% - #{$wuiSizeS});
    margin-right: $wuiSizeS;
    margin-bottom: 0;
  }

  .wuiFormRow__fieldWrapper {
    width: 67%;
  }

  + .wuiFormRow--horizontal {
    margin-top: $wuiSizeS;
  }

  + .wuiFormRow--horizontal.wuiFormRow--hasSwitch {
    margin-top: $wuiSizeM; // More spacing since we reduced the height to match that of the switch
  }

  &.wuiFormRow--hasSwitch {
    .wuiFormRow__labelWrapper {
      line-height: $wuiSwitchHeight - 1px; // The 1px less helps the alignment of the text baseline
      width: auto;
      min-width: calc(33% - #{$wuiSizeS});
    }

    .wuiFormRow__fieldWrapper {
      width: auto;

      .wuiSwitch--compressed {
        margin-top: $wuiSizeXS / 2; // Better vertical alignment of a compressed switch to the horizontal label
      }
    }

    + .wuiFormRow--horizontal {
      margin-top: $wuiSizeM; // More spacing since we reduced the height to match that of the switch
    }
  }
}

.wuiFormRow__fieldWrapperDisplayOnly {
  min-height: $wuiFormControlHeight;
  display: flex;
  align-items: center;
}

.wuiFormRow--compressed {
  &.wuiFormRow--hasEmptyLabelSpace {
    min-height: $wuiFormControlCompressedHeight;
  }

  .wuiFormRow__fieldWrapperDisplayOnly {
    min-height: $wuiFormControlCompressedHeight;
  }
}
