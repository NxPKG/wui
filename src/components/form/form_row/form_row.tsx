/*
 * Copyright 2022 Wazuh Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * NOTICE: THIS FILE HAS BEEN MODIFIED BY WAZUH INC UNDER COMPLIANCE WITH THE APACHE 2.0 LICENSE FROM THE ORIGINAL WORK
 * OF THE COMPANY Elasticsearch B.V.
 *
 * THE FOLLOWING IS THE COPYRIGHT OF THE ORIGINAL DOCUMENT:
 *
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import React, {
  cloneElement,
  Component,
  Children,
  HTMLAttributes,
  ReactElement,
  ReactNode,
} from 'react';
import classNames from 'classnames';
import { ExclusiveUnion, CommonProps, keysOf } from '../../common';

import { get } from '../../../services/objects';

import { WuiFormHelpText } from '../form_help_text';
import { WuiFormErrorText } from '../form_error_text';
import { WuiFormLabel } from '../form_label';

import { htmlIdGenerator } from '../../../services/accessibility';

const displayToClassNameMap = {
  row: null,
  rowCompressed: 'wuiFormRow--compressed',
  columnCompressed: 'wuiFormRow--compressed wuiFormRow--horizontal',
  center: null,
  centerCompressed: 'wuiFormRow--compressed',
  columnCompressedSwitch:
    'wuiFormRow--compressed wuiFormRow--horizontal wuiFormRow--hasSwitch',
};

export const DISPLAYS = keysOf(displayToClassNameMap);

export type WuiFormRowDisplayKeys = keyof typeof displayToClassNameMap;

interface WuiFormRowState {
  isFocused: boolean;
  id: string;
}

type WuiFormRowCommonProps = CommonProps & {
  /**
   * When `rowCompressed`, just tightens up the spacing;
   * Set to `columnCompressed` if compressed
   * and horizontal layout is needed.
   * Set to `center` or `centerCompressed` to align non-input
   * content better with inline rows.
   * Set to `columnCompressedSwitch` if the form control being passed
   * as the child is a switch.
   */
  display?: WuiFormRowDisplayKeys;
  hasEmptyLabelSpace?: boolean;
  fullWidth?: boolean;
  /**
   * IDs of additional elements that should be part of children's `aria-describedby`
   */
  describedByIds?: string[];
  /**
   * Escape hatch to not render duplicate labels if the child also renders a label
   */
  hasChildLabel?: boolean;
  /**
   * ReactElement to render as this component's content
   */
  children: ReactElement;
  label?: ReactNode;
  /**
   * Adds an extra node to the right of the form label without
   * being contained inside the form label. Good for things
   * like documentation links.
   */
  labelAppend?: any;
  id?: string;
  isInvalid?: boolean;
  error?: ReactNode | ReactNode[];
  helpText?: ReactNode;
  /**
   * **DEPRECATED: use `display: rowCompressed` instead.**
   * When `true`, tightens up the spacing.
   */
  compressed?: boolean;
  /**
   * **DEPRECATED: use `display: center` instead.**
   * Vertically centers non-input style content so it aligns
   * better with input style content.
   */
  displayOnly?: boolean;
};

type LabelProps = {
  labelType?: 'label';
} & WuiFormRowCommonProps &
  HTMLAttributes<HTMLDivElement>;

type LegendProps = {
  /**
   * Defaults to rendering a `<label>` but if passed `'legend'` for labelType,
   * will render both a `<legend>` and the surrounding container as a `<fieldset>`
   */
  labelType?: 'legend';
} & WuiFormRowCommonProps &
  HTMLAttributes<HTMLFieldSetElement>;

export type WuiFormRowProps = ExclusiveUnion<LabelProps, LegendProps>;

export class WuiFormRow extends Component<WuiFormRowProps, WuiFormRowState> {
  static defaultProps = {
    display: 'row',
    hasEmptyLabelSpace: false,
    fullWidth: false,
    describedByIds: [],
    labelType: 'label',
    hasChildLabel: true,
  };

  state: WuiFormRowState = {
    isFocused: false,
    id: this.props.id || htmlIdGenerator()(),
  };

  onFocus = (...args: any[]) => {
    // Doing this to allow onFocus to be called correctly from the child input element as this component overrides it
    const onChildFocus = get(this.props, 'children.props.onFocus');
    if (onChildFocus) {
      onChildFocus(...args);
    }

    this.setState(({ isFocused }) => {
      if (!isFocused) {
        return {
          isFocused: true,
        };
      } else {
        return null;
      }
    });
  };

  onBlur = (...args: any[]) => {
    // Doing this to allow onBlur to be called correctly from the child input element as this component overrides it
    const onChildBlur = get(this.props, 'children.props.onBlur');
    if (onChildBlur) {
      onChildBlur(...args);
    }

    this.setState({
      isFocused: false,
    });
  };

  render() {
    const {
      children,
      helpText,
      isInvalid,
      error,
      label,
      labelType,
      labelAppend,
      hasEmptyLabelSpace,
      fullWidth,
      className,
      describedByIds,
      compressed,
      display,
      displayOnly,
      hasChildLabel,
      id: propsId,
      ...rest
    } = this.props;

    const { id } = this.state;

    /**
     * Remove when `compressed` is deprecated
     */
    let shimDisplay: WuiFormRowDisplayKeys;
    if (compressed && display === 'row') {
      shimDisplay = 'rowCompressed';
    } else {
      /**
       * Safe use of ! as prop default is 'row'
       */
      shimDisplay = display!;
    }

    /**
     * Remove when `displayOnly` is deprecated
     */
    if (compressed && displayOnly) {
      shimDisplay = 'centerCompressed';
    } else if (displayOnly && display === 'row') {
      shimDisplay = 'center';
    }

    const classes = classNames(
      'wuiFormRow',
      {
        'wuiFormRow--hasEmptyLabelSpace': hasEmptyLabelSpace,
        'wuiFormRow--fullWidth': fullWidth,
      },
      displayToClassNameMap[shimDisplay],
      className
    );

    let optionalHelpText;

    if (helpText) {
      optionalHelpText = (
        <WuiFormHelpText id={`${id}-help`} className="wuiFormRow__text">
          {helpText}
        </WuiFormHelpText>
      );
    }

    let optionalErrors;

    if (error && isInvalid) {
      const errorTexts = Array.isArray(error) ? error : [error];
      optionalErrors = errorTexts.map((error, i) => {
        const key = typeof error === 'string' ? error : i;
        return (
          <WuiFormErrorText
            key={key}
            id={`${id}-error-${i}`}
            className="wuiFormRow__text">
            {error}
          </WuiFormErrorText>
        );
      });
    }

    let optionalLabel;
    const isLegend = label && labelType === 'legend' ? true : false;

    if (label || labelAppend) {
      let labelProps = {};
      if (isLegend) {
        labelProps = {
          type: labelType,
        };
      } else {
        labelProps = {
          htmlFor: hasChildLabel ? id : undefined,
          isFocused: this.state.isFocused,
          type: labelType,
        };
      }
      optionalLabel = (
        <div className="wuiFormRow__labelWrapper">
          <WuiFormLabel
            className="wuiFormRow__label"
            isInvalid={isInvalid}
            aria-invalid={isInvalid}
            {...labelProps}>
            {label}
          </WuiFormLabel>
          {labelAppend && ' '}
          {labelAppend}
        </div>
      );
    }

    const optionalProps: React.AriaAttributes = {};
    /**
     * Safe use of ! as default prop is []
     */
    const describingIds = [...describedByIds!];

    if (optionalHelpText) {
      describingIds.push(optionalHelpText.props.id);
    }

    if (optionalErrors) {
      optionalErrors.forEach(error => describingIds.push(error.props.id));
    }

    if (describingIds.length > 0) {
      optionalProps['aria-describedby'] = describingIds.join(' ');
    }

    const field = cloneElement(Children.only(children), {
      id,
      onFocus: this.onFocus,
      onBlur: this.onBlur,
      ...optionalProps,
    });

    const fieldWrapperClasses = classNames('wuiFormRow__fieldWrapper', {
      wuiFormRow__fieldWrapperDisplayOnly:
        /**
         * Safe use of ! as default prop is 'row'
         */
        displayOnly || display!.startsWith('center'),
    });

    const sharedProps = {
      className: classes,
      id: `${id}-row`,
    };

    const contents = (
      <React.Fragment>
        {optionalLabel}
        <div className={fieldWrapperClasses}>
          {field}
          {optionalErrors}
          {optionalHelpText}
        </div>
      </React.Fragment>
    );

    return labelType === 'legend' ? (
      <fieldset
        {...sharedProps}
        {...(rest as HTMLAttributes<HTMLFieldSetElement>)}>
        {contents}
      </fieldset>
    ) : (
      <div {...sharedProps} {...(rest as HTMLAttributes<HTMLDivElement>)}>
        {contents}
      </div>
    );
  }
}
