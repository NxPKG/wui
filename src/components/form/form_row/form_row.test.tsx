/*
 * Copyright 2022 Wazuh Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * NOTICE: THIS FILE HAS BEEN MODIFIED BY WAZUH INC UNDER COMPLIANCE WITH THE APACHE 2.0 LICENSE FROM THE ORIGINAL WORK
 * OF THE COMPANY Elasticsearch B.V.
 *
 * THE FOLLOWING IS THE COPYRIGHT OF THE ORIGINAL DOCUMENT:
 *
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import React from 'react';
import { shallow, render, mount } from 'enzyme';
import { requiredProps } from '../../../test';

import { WuiFormRow, DISPLAYS } from './form_row';

jest.mock('./../../../services/accessibility', () => ({
  htmlIdGenerator: () => () => 'generated-id',
}));

describe('WuiFormRow', () => {
  test('is rendered', () => {
    const component = render(
      <WuiFormRow {...requiredProps}>
        <input />
      </WuiFormRow>
    );

    expect(component).toMatchSnapshot();
  });

  test('ties together parts for accessibility', () => {
    const props = {
      label: 'Label',
      helpText: 'Help text',
      isInvalid: true,
      error: ['Error one', 'Error two'],
    };

    const tree = shallow(
      <WuiFormRow {...requiredProps} {...props}>
        <input />
      </WuiFormRow>
    );

    // Input is labeled by the label.
    expect(tree.find('input').prop('id')).toEqual('generated-id');
    expect(tree.find('WuiFormLabel').prop('htmlFor')).toEqual('generated-id');

    // Input is described by help and error text.
    expect(tree.find('WuiFormHelpText').prop('id')).toEqual(
      'generated-id-help'
    );
    expect(
      tree
        .find('WuiFormErrorText')
        .at(0)
        .prop('id')
    ).toEqual('generated-id-error-0');
    expect(
      tree
        .find('WuiFormErrorText')
        .at(1)
        .prop('id')
    ).toEqual('generated-id-error-1');
    expect(tree.find('input').prop('aria-describedby')).toEqual(
      'generated-id-help generated-id-error-0 generated-id-error-1'
    );
  });

  describe('props', () => {
    test('label is rendered', () => {
      const component = shallow(
        <WuiFormRow label="label">
          <input />
        </WuiFormRow>
      );

      expect(component).toMatchSnapshot();
    });

    test('label append is rendered', () => {
      const component = shallow(
        <WuiFormRow label="label" labelAppend="append">
          <input />
        </WuiFormRow>
      );

      expect(component).toMatchSnapshot();
    });

    test('label renders as a legend and subsquently a fieldset wrapper', () => {
      const component = shallow(
        <WuiFormRow label="label" labelType="legend">
          <input />
        </WuiFormRow>
      );

      expect(component).toMatchSnapshot();
    });

    test('describedByIds is rendered', () => {
      const component = shallow(
        <WuiFormRow describedByIds={['generated-id-additional']}>
          <input />
        </WuiFormRow>
      );

      expect(component).toMatchSnapshot();
    });

    test('id is rendered', () => {
      const component = render(
        <WuiFormRow id="id">
          <input />
        </WuiFormRow>
      );

      expect(component).toMatchSnapshot();
    });

    test('isInvalid is rendered', () => {
      const component = render(
        <WuiFormRow isInvalid label="label">
          <input />
        </WuiFormRow>
      );

      expect(component).toMatchSnapshot();
    });

    test('error as string is rendered', () => {
      const component = render(
        <WuiFormRow error="Error" isInvalid={true}>
          <input />
        </WuiFormRow>
      );

      expect(component).toMatchSnapshot();
    });

    test('error as array is rendered', () => {
      const component = render(
        <WuiFormRow error={['Error', 'Error2']} isInvalid={true}>
          <input />
        </WuiFormRow>
      );

      expect(component).toMatchSnapshot();
    });

    test('error is not rendered if isInvalid is false', () => {
      const component = render(
        <WuiFormRow error={['Error']} isInvalid={false}>
          <input />
        </WuiFormRow>
      );

      expect(component).toMatchSnapshot();
    });

    test('helpText is rendered', () => {
      const component = render(
        <WuiFormRow helpText={<span>This is help text.</span>}>
          <input />
        </WuiFormRow>
      );

      expect(component).toMatchSnapshot();
    });

    test('hasEmptyLabelSpace is rendered', () => {
      const component = render(
        <WuiFormRow hasEmptyLabelSpace>
          <input />
        </WuiFormRow>
      );

      expect(component).toMatchSnapshot();
    });

    test('fullWidth is rendered', () => {
      const component = render(
        <WuiFormRow fullWidth>
          <input />
        </WuiFormRow>
      );

      expect(component).toMatchSnapshot();
    });

    test('displayOnly is rendered', () => {
      const component = render(
        <WuiFormRow displayOnly>
          <span>just some text</span>
        </WuiFormRow>
      );

      expect(component).toMatchSnapshot();
    });

    describe('compressed', () => {
      test('is rendered', () => {
        const component = render(
          <WuiFormRow compressed>
            <input />
          </WuiFormRow>
        );

        expect(component).toMatchSnapshot();
      });
    });

    describe('display type', () => {
      DISPLAYS.forEach(display => {
        test(`${display} is rendered`, () => {
          const component = render(
            <WuiFormRow display={display}>
              <input />
            </WuiFormRow>
          );

          expect(component).toMatchSnapshot();
        });
      });
    });
  });

  describe('behavior', () => {
    describe('onFocus', () => {
      test('is called in child', () => {
        const focusMock = jest.fn();

        const component = mount(
          <WuiFormRow label={<span>Label</span>}>
            <input onFocus={focusMock} />
          </WuiFormRow>
        );

        component.find('input').simulate('focus');

        expect(focusMock).toBeCalledTimes(1);

        // Ensure the focus event is properly fired on the parent
        // which will pass down to the WuiFormLabel
        expect(component).toMatchSnapshot();
      });

      test('works in parent even if not in child', () => {
        const component = mount(
          <WuiFormRow label={<span>Label</span>}>
            <input />
          </WuiFormRow>
        );

        component.find('input').simulate('focus');

        // Ensure the focus event is properly fired on the parent
        // which will pass down to the WuiFormLabel
        expect(component).toMatchSnapshot();
      });
    });

    describe('onBlur', () => {
      test('is called in child', () => {
        const blurMock = jest.fn();

        const component = mount(
          <WuiFormRow label={<span>Label</span>}>
            <input onBlur={blurMock} />
          </WuiFormRow>
        );

        component.find('input').simulate('blur');

        expect(blurMock).toBeCalledTimes(1);

        // Ensure the blur event is properly fired on the parent
        // which will pass down to the WuiFormLabel
        expect(component).toMatchSnapshot();
      });

      test('works in parent even if not in child', () => {
        const component = mount(
          <WuiFormRow label={<span>Label</span>}>
            <input />
          </WuiFormRow>
        );

        component.find('input').simulate('blur');

        // Ensure the blur event is properly fired on the parent
        // which will pass down to the WuiFormLabel
        expect(component).toMatchSnapshot();
      });
    });
  });
});
