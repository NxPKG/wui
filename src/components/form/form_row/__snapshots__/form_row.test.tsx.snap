// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WuiFormRow behavior onBlur is called in child 1`] = `
<WuiFormRow
  describedByIds={Array []}
  display="row"
  fullWidth={false}
  hasChildLabel={true}
  hasEmptyLabelSpace={false}
  label={
    <span>
      Label
    </span>
  }
  labelType="label"
>
  <div
    className="wuiFormRow"
    id="generated-id-row"
  >
    <div
      className="wuiFormRow__labelWrapper"
    >
      <WuiFormLabel
        className="wuiFormRow__label"
        htmlFor="generated-id"
        isFocused={false}
        type="label"
      >
        <label
          className="wuiFormLabel wuiFormRow__label"
          htmlFor="generated-id"
        >
          <span>
            Label
          </span>
        </label>
      </WuiFormLabel>
    </div>
    <div
      className="wuiFormRow__fieldWrapper"
    >
      <input
        id="generated-id"
        onBlur={[Function]}
        onFocus={[Function]}
      />
    </div>
  </div>
</WuiFormRow>
`;

exports[`WuiFormRow behavior onBlur works in parent even if not in child 1`] = `
<WuiFormRow
  describedByIds={Array []}
  display="row"
  fullWidth={false}
  hasChildLabel={true}
  hasEmptyLabelSpace={false}
  label={
    <span>
      Label
    </span>
  }
  labelType="label"
>
  <div
    className="wuiFormRow"
    id="generated-id-row"
  >
    <div
      className="wuiFormRow__labelWrapper"
    >
      <WuiFormLabel
        className="wuiFormRow__label"
        htmlFor="generated-id"
        isFocused={false}
        type="label"
      >
        <label
          className="wuiFormLabel wuiFormRow__label"
          htmlFor="generated-id"
        >
          <span>
            Label
          </span>
        </label>
      </WuiFormLabel>
    </div>
    <div
      className="wuiFormRow__fieldWrapper"
    >
      <input
        id="generated-id"
        onBlur={[Function]}
        onFocus={[Function]}
      />
    </div>
  </div>
</WuiFormRow>
`;

exports[`WuiFormRow behavior onFocus is called in child 1`] = `
<WuiFormRow
  describedByIds={Array []}
  display="row"
  fullWidth={false}
  hasChildLabel={true}
  hasEmptyLabelSpace={false}
  label={
    <span>
      Label
    </span>
  }
  labelType="label"
>
  <div
    className="wuiFormRow"
    id="generated-id-row"
  >
    <div
      className="wuiFormRow__labelWrapper"
    >
      <WuiFormLabel
        className="wuiFormRow__label"
        htmlFor="generated-id"
        isFocused={true}
        type="label"
      >
        <label
          className="wuiFormLabel wuiFormRow__label wuiFormLabel-isFocused"
          htmlFor="generated-id"
        >
          <span>
            Label
          </span>
        </label>
      </WuiFormLabel>
    </div>
    <div
      className="wuiFormRow__fieldWrapper"
    >
      <input
        id="generated-id"
        onBlur={[Function]}
        onFocus={[Function]}
      />
    </div>
  </div>
</WuiFormRow>
`;

exports[`WuiFormRow behavior onFocus works in parent even if not in child 1`] = `
<WuiFormRow
  describedByIds={Array []}
  display="row"
  fullWidth={false}
  hasChildLabel={true}
  hasEmptyLabelSpace={false}
  label={
    <span>
      Label
    </span>
  }
  labelType="label"
>
  <div
    className="wuiFormRow"
    id="generated-id-row"
  >
    <div
      className="wuiFormRow__labelWrapper"
    >
      <WuiFormLabel
        className="wuiFormRow__label"
        htmlFor="generated-id"
        isFocused={true}
        type="label"
      >
        <label
          className="wuiFormLabel wuiFormRow__label wuiFormLabel-isFocused"
          htmlFor="generated-id"
        >
          <span>
            Label
          </span>
        </label>
      </WuiFormLabel>
    </div>
    <div
      className="wuiFormRow__fieldWrapper"
    >
      <input
        id="generated-id"
        onBlur={[Function]}
        onFocus={[Function]}
      />
    </div>
  </div>
</WuiFormRow>
`;

exports[`WuiFormRow is rendered 1`] = `
<div
  aria-label="aria-label"
  class="wuiFormRow testClass1 testClass2"
  data-test-subj="test subject string"
  id="generated-id-row"
>
  <div
    class="wuiFormRow__fieldWrapper"
  >
    <input
      id="generated-id"
    />
  </div>
</div>
`;

exports[`WuiFormRow props compressed is rendered 1`] = `
<div
  class="wuiFormRow wuiFormRow--compressed"
  id="generated-id-row"
>
  <div
    class="wuiFormRow__fieldWrapper"
  >
    <input
      id="generated-id"
    />
  </div>
</div>
`;

exports[`WuiFormRow props describedByIds is rendered 1`] = `
<div
  className="wuiFormRow"
  id="generated-id-row"
>
  <div
    className="wuiFormRow__fieldWrapper"
  >
    <input
      aria-describedby="generated-id-additional"
      id="generated-id"
      onBlur={[Function]}
      onFocus={[Function]}
    />
  </div>
</div>
`;

exports[`WuiFormRow props display type center is rendered 1`] = `
<div
  class="wuiFormRow"
  id="generated-id-row"
>
  <div
    class="wuiFormRow__fieldWrapper wuiFormRow__fieldWrapperDisplayOnly"
  >
    <input
      id="generated-id"
    />
  </div>
</div>
`;

exports[`WuiFormRow props display type centerCompressed is rendered 1`] = `
<div
  class="wuiFormRow wuiFormRow--compressed"
  id="generated-id-row"
>
  <div
    class="wuiFormRow__fieldWrapper wuiFormRow__fieldWrapperDisplayOnly"
  >
    <input
      id="generated-id"
    />
  </div>
</div>
`;

exports[`WuiFormRow props display type columnCompressed is rendered 1`] = `
<div
  class="wuiFormRow wuiFormRow--compressed wuiFormRow--horizontal"
  id="generated-id-row"
>
  <div
    class="wuiFormRow__fieldWrapper"
  >
    <input
      id="generated-id"
    />
  </div>
</div>
`;

exports[`WuiFormRow props display type columnCompressedSwitch is rendered 1`] = `
<div
  class="wuiFormRow wuiFormRow--compressed wuiFormRow--horizontal wuiFormRow--hasSwitch"
  id="generated-id-row"
>
  <div
    class="wuiFormRow__fieldWrapper"
  >
    <input
      id="generated-id"
    />
  </div>
</div>
`;

exports[`WuiFormRow props display type row is rendered 1`] = `
<div
  class="wuiFormRow"
  id="generated-id-row"
>
  <div
    class="wuiFormRow__fieldWrapper"
  >
    <input
      id="generated-id"
    />
  </div>
</div>
`;

exports[`WuiFormRow props display type rowCompressed is rendered 1`] = `
<div
  class="wuiFormRow wuiFormRow--compressed"
  id="generated-id-row"
>
  <div
    class="wuiFormRow__fieldWrapper"
  >
    <input
      id="generated-id"
    />
  </div>
</div>
`;

exports[`WuiFormRow props displayOnly is rendered 1`] = `
<div
  class="wuiFormRow"
  id="generated-id-row"
>
  <div
    class="wuiFormRow__fieldWrapper wuiFormRow__fieldWrapperDisplayOnly"
  >
    <span
      id="generated-id"
    >
      just some text
    </span>
  </div>
</div>
`;

exports[`WuiFormRow props error as array is rendered 1`] = `
<div
  class="wuiFormRow"
  id="generated-id-row"
>
  <div
    class="wuiFormRow__fieldWrapper"
  >
    <input
      aria-describedby="generated-id-error-0 generated-id-error-1"
      id="generated-id"
    />
    <div
      aria-live="polite"
      class="wuiFormErrorText wuiFormRow__text"
      id="generated-id-error-0"
    >
      Error
    </div>
    <div
      aria-live="polite"
      class="wuiFormErrorText wuiFormRow__text"
      id="generated-id-error-1"
    >
      Error2
    </div>
  </div>
</div>
`;

exports[`WuiFormRow props error as string is rendered 1`] = `
<div
  class="wuiFormRow"
  id="generated-id-row"
>
  <div
    class="wuiFormRow__fieldWrapper"
  >
    <input
      aria-describedby="generated-id-error-0"
      id="generated-id"
    />
    <div
      aria-live="polite"
      class="wuiFormErrorText wuiFormRow__text"
      id="generated-id-error-0"
    >
      Error
    </div>
  </div>
</div>
`;

exports[`WuiFormRow props error is not rendered if isInvalid is false 1`] = `
<div
  class="wuiFormRow"
  id="generated-id-row"
>
  <div
    class="wuiFormRow__fieldWrapper"
  >
    <input
      id="generated-id"
    />
  </div>
</div>
`;

exports[`WuiFormRow props fullWidth is rendered 1`] = `
<div
  class="wuiFormRow wuiFormRow--fullWidth"
  id="generated-id-row"
>
  <div
    class="wuiFormRow__fieldWrapper"
  >
    <input
      id="generated-id"
    />
  </div>
</div>
`;

exports[`WuiFormRow props hasEmptyLabelSpace is rendered 1`] = `
<div
  class="wuiFormRow wuiFormRow--hasEmptyLabelSpace"
  id="generated-id-row"
>
  <div
    class="wuiFormRow__fieldWrapper"
  >
    <input
      id="generated-id"
    />
  </div>
</div>
`;

exports[`WuiFormRow props helpText is rendered 1`] = `
<div
  class="wuiFormRow"
  id="generated-id-row"
>
  <div
    class="wuiFormRow__fieldWrapper"
  >
    <input
      aria-describedby="generated-id-help"
      id="generated-id"
    />
    <div
      class="wuiFormHelpText wuiFormRow__text"
      id="generated-id-help"
    >
      <span>
        This is help text.
      </span>
    </div>
  </div>
</div>
`;

exports[`WuiFormRow props id is rendered 1`] = `
<div
  class="wuiFormRow"
  id="id-row"
>
  <div
    class="wuiFormRow__fieldWrapper"
  >
    <input
      id="id"
    />
  </div>
</div>
`;

exports[`WuiFormRow props isInvalid is rendered 1`] = `
<div
  class="wuiFormRow"
  id="generated-id-row"
>
  <div
    class="wuiFormRow__labelWrapper"
  >
    <label
      aria-invalid="true"
      class="wuiFormLabel wuiFormRow__label wuiFormLabel-isInvalid"
      for="generated-id"
    >
      label
    </label>
  </div>
  <div
    class="wuiFormRow__fieldWrapper"
  >
    <input
      id="generated-id"
    />
  </div>
</div>
`;

exports[`WuiFormRow props label append is rendered 1`] = `
<div
  className="wuiFormRow"
  id="generated-id-row"
>
  <div
    className="wuiFormRow__labelWrapper"
  >
    <WuiFormLabel
      className="wuiFormRow__label"
      htmlFor="generated-id"
      isFocused={false}
      type="label"
    >
      label
    </WuiFormLabel>
     
    append
  </div>
  <div
    className="wuiFormRow__fieldWrapper"
  >
    <input
      id="generated-id"
      onBlur={[Function]}
      onFocus={[Function]}
    />
  </div>
</div>
`;

exports[`WuiFormRow props label is rendered 1`] = `
<div
  className="wuiFormRow"
  id="generated-id-row"
>
  <div
    className="wuiFormRow__labelWrapper"
  >
    <WuiFormLabel
      className="wuiFormRow__label"
      htmlFor="generated-id"
      isFocused={false}
      type="label"
    >
      label
    </WuiFormLabel>
  </div>
  <div
    className="wuiFormRow__fieldWrapper"
  >
    <input
      id="generated-id"
      onBlur={[Function]}
      onFocus={[Function]}
    />
  </div>
</div>
`;

exports[`WuiFormRow props label renders as a legend and subsquently a fieldset wrapper 1`] = `
<fieldset
  className="wuiFormRow"
  id="generated-id-row"
>
  <div
    className="wuiFormRow__labelWrapper"
  >
    <WuiFormLabel
      className="wuiFormRow__label"
      type="legend"
    >
      label
    </WuiFormLabel>
  </div>
  <div
    className="wuiFormRow__fieldWrapper"
  >
    <input
      id="generated-id"
      onBlur={[Function]}
      onFocus={[Function]}
    />
  </div>
</fieldset>
`;
